core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -187,7 +187,7 @@ class TransactionCoordinator(brokerId: Int,
     } else {
       // caller should have synchronized on txnMetadata already
       txnMetadata.state match {
-        case PrepareAbort | PrepareCommit =>
+        case PrepareAbort | PrepareCommit | Dead =>
```

Comment:
The point is that, actually we never "complete the transition to Dead" in the callback, but just remove it from the cache; so `Dead` should only be seen on the pending state field, but never be seen on the state field.